kani-driver/src/project.rs

```diff
@@ -0,0 +1,282 @@
+/// the same type are linked as part of creating the project. Except for `symtab` files used for
+/// demangling `C` when on legacy linker mode.
+///
+/// However, one artifact can be used for multiple harnesses. This will depend on the type of
+/// artifact, but it should be transparent for the user of this object.
+#[derive(Debug, Default)]
+pub struct Project {
+    /// Each target crate metadata.
+    pub metadata: Vec<KaniMetadata>,
+    /// The directory where all outputs should be directed to. This path represents the canonical
+    /// version of outdir.
+    pub outdir: PathBuf,
+    /// The collection of artifacts kept as part of this project.
+    artifacts: Vec<Artifact>,
+    /// A flag that allow us to provide a consistent behavior for `--legacy-linker`.
+    /// If this flag is true, there should only be up to one artifact of a
+    merged_artifacts: bool,
+}
+
+impl Project {
+    /// Get all harnesses from a project. This will include all test and proof harnesses.
+    /// We could create a `get_proof_harnesses` and a `get_tests_harnesses` later if we see the
+    /// need to split them.
+    pub fn get_all_harnesses(&self) -> Vec<&HarnessMetadata> {
+        self.metadata
+            .iter()
+            .flat_map(|crate_metadata| {
+                crate_metadata.proof_harnesses.iter().chain(crate_metadata.test_harnesses.iter())
+            })
+            .collect()
+    }
+
+    /// Return the matching artifact for the given harness.
+    ///
+    /// If the harness has information about the goto_file we can use that to find the exact file.
+    /// For cases where there is no goto_file, we just assume that everything has been linked
+    /// together. I.e.: There should only be one artifact of the given type.
+    pub fn get_harness_artifact(
+        &self,
+        harness: &HarnessMetadata,
+        typ: ArtifactType,
+    ) -> Option<&Artifact> {
+        let expected_path = if self.merged_artifacts {
+            None
+        } else {
+            harness.goto_file.as_ref().map(|goto_file| convert_type(goto_file, SymTabGoto, typ))
+        };
+        trace!(?harness.goto_file, ?expected_path, ?typ, "get_harness_artifact");
+        self.artifacts.iter().find(|artifact| {
+            artifact.has_type(typ)
+                && expected_path.as_ref().map_or(true, |goto_file| *goto_file == artifact.path)
+        })
+    }
+}
+
+/// Information about a build artifact.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct Artifact {
+    /// The path for this artifact in the canonical form.
+    path: PathBuf,
+    /// The type of artifact.
+    typ: ArtifactType,
+}
+
+impl AsRef<Path> for Artifact {
+    fn as_ref(&self) -> &Path {
+        self.path.as_ref()
+    }
+}
+
+impl Deref for Artifact {
+    type Target = Path;
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl Artifact {
+    /// Create a new existing artifact.
```

Comment:
"new" and "existing" sound a bit conflicting. Perhaps we can say "Create an artifact object for an artifact whose file exists on disk"?